### DISCUSSION TITLE
User Input: Ctrl + V a column from excel or similar

### DISCUSSION THREAD

**Discussion by thickmn** [zauberzeug/nicegui#1398] [Ideas / Feature Requests]
### Question

Is there any functionality that allows a user to ctrl+v paste a column from an existing excel or csv column into a UI table?  Similar to how columns can be pasted into an excel table. 

![github question](https://github.com/zauberzeug/nicegui/assets/109317955/52abbc58-9e94-45ad-bb67-b3c2b30b1eb7)

**Comment by rodja** [comment]
I'm not aware of it, so I'll convert this to a feature request. Anyone interested in looking into this?

**Reply by thickmn** [reply]
I suppose a textarea and a whitespace delimiter could be used to accept a user copy/pasting the values from an excel column. Then clearing the textarea container and rendering an aggrid table in its place. 

I thought about this use case a little more, and I'd be happy with a textarea input that is styled with an outline to visually represent a column. Perhaps it could stretch with the length of the input?

**Reply by thickmn** [reply]
Ultimately, though -- it would be pretty cool to have a multi line input directly into an aggrid column

**Comment by falkoschindler** [comment]
Maybe this could be a starting point?

```py
async def paste():
    clipboard = await ui.run_javascript('return navigator.clipboard.readText()')
    for row in clipboard.split('\n'):
        name, age = row.split('\t')
        grid.options['rowData'].append({'name': name, 'age': int(age)})
    grid.update()

grid = ui.aggrid({
    'defaultColDef': {'flex': 1},
    'columnDefs': [
        {'headerName': 'Name', 'field': 'name'},
        {'headerName': 'Age', 'field': 'age'},
    ],
    'rowData': [
        {'name': 'Alice', 'age': 18},
        {'name': 'Bob', 'age': 21},
        {'name': 'Carol', 'age': 42},
    ],
}).on('keydown.meta.v', paste)
```

Note that this listens for Meta+V (i.e. Command) on Mac. For Windows you might want to use `.ctrl` instead.

**Reply by thickmn** [reply]
Ah, slick! I'll try when I get to my desk in the morning

**Comment by thickmn** [comment]
> Maybe this could be a starting point?
> 
> 
> 
> ```py
> 
> async def paste():
> 
>     clipboard = await ui.run_javascript('return navigator.clipboard.readText()')
> 
>     for row in clipboard.split('\n'):
> 
>         name, age = row.split('\t')
> 
>         grid.options['rowData'].append({'name': name, 'age': int(age)})
> 
>     grid.update()
> 
> 
> 
> grid = ui.aggrid({
> 
>     'defaultColDef': {'flex': 1},
> 
>     'columnDefs': [
> 
>         {'headerName': 'Name', 'field': 'name'},
> 
>         {'headerName': 'Age', 'field': 'age'},
> 
>     ],
> 
>     'rowData': [
> 
>         {'name': 'Alice', 'age': 18},
> 
>         {'name': 'Bob', 'age': 21},
> 
>         {'name': 'Carol', 'age': 42},
> 
>     ],
> 
> }).on('keydown.meta.v', paste)
> 
> ```
> 
> 
> 
> Note that this listens for Meta+V (i.e. Command) on Mac. For Windows you might want to use `.ctrl` instead.

This worked well for a use case where a user needed functionality to paste a list of IDs to be excluded from the results of a sql query. I kept a single column header and an empty row in rowData and added {editable: True} to columnDefs. A submit button and a clear button to submit or remove from an embedded SQLite database. 

Is the best way to populate rowData from a SQLAlchemy session to pass a sorted list of dicts to rowData? 

Also, we can close this as answered, and a continued thanks to the team for your work on such a well managed project in nicegui!

**Comment by apsr3112** [comment]
Hi
could you help me?
I have an application that allows to display tables as an excel file and sometimes, the data is displayed combining text and shapes, as shown in the attached image.
![picture_1](https://github.com/zauberzeug/nicegui/assets/158159276/c30dbb2e-8228-4684-a46f-b4da27306086)

Would it be possible to adapt my grid to AG Grid, and be able to copy cells and rows, using Ctrl+c (p.e.) and keeping the format that is displayed?
